File: day04/solution.jl
function parsePairs(pair)
  (elf1, elf2) = split(pair, ",")
  return (parseRange(elf1), parseRange(elf2))
end

function parseRange(elf)
  (rStart, rEnd) = split(elf, "-")
  return parse(Int64, rStart):parse(Int64, rEnd)
end

open("input", "r") do file
  countSubset = 0
  anyOverlap = 0
  while !eof(file)
    line = readline(file)
    (elf1, elf2) = parsePairs(line)
    common = intersect(elf1, elf2)
    if elf1 == common || elf2 == common
      countSubset += 1
    end
    if !isempty(common)
      anyOverlap += 1
    end
  end
  println("subset count: $countSubset")
  println("any overlap count: $anyOverlap")
end;

File: day04/solution.exs
defmodule Functions do
  def parse_pairs([elf1, elf2]), do: {parse_range(elf1), parse_range(elf2)}

  defp parse_range(range) do
    [r_start, r_end] = String.split(range, "-")
    String.to_integer(r_start)..String.to_integer(r_end)
  end

  @doc """
  Returns true if one range is wholly inside the other
  """
  def overlap?({elf1, elf2}) do
    Enum.all?(elf1, fn x -> x in elf2 end) || Enum.all?(elf2, fn x -> x in elf1 end)
  end
end

pairs = File.stream!("input", [:utf8], :line)
|> Stream.map(&String.trim/1)
|> Stream.map(&(String.split(&1, ",")))
|> Enum.map(&Functions.parse_pairs/1)

# Part 1 - Count of pairs where one pair is wholly inside the other
pairs
|> Enum.count(&Functions.overlap?/1)
|> IO.inspect(label: "Count of whole overlap")

# Part 2 - Count of pairs with ANY overlap (or things that DON'T have NO OVERLAP)
pairs
|> Enum.reject(fn {elf1, elf2} -> Range.disjoint?(elf1, elf2) end)
|> Enum.count()
|> IO.inspect(label: "Count of any overlap")
